### ISSUE TITLE
upgrade fails with: spec.users.passwd in body must be of type string: "null"

### ISSUE THREAD

**Issue by andrewsykim** [vmware/cluster-api-upgrade-tool#135]
TL;Dr I think this is because the bootstrap API doesn't have `omitempty` for `spec.users.passwd`. I can open an issue there if that's the case, but it's not clear to me yet because it seems like CAPA doesn't set `users.spec.passwd` either but it doesn't seem to be an issue there.

From the upgrade tool using CAPV w/ SSH keys for user access:
```
INFO[0000] Updating machines
INFO[0000] Running etcdctl                               args="member list -w json" pod=etcd-target-cluster02-controlplane-0
INFO[0000] etcdctl stdout: {"header":{"cluster_id":9521744471245018578,"member_id":7742574140227613697,"raft_term":2},"members":[{"ID":7742574140227613697,"name":"target-cluster02-controlplane-0","peerURLs":["https://192.168.3.98:2380"],"clientURLs":["https://192.168.3.98:2379"]}]}
INFO[0000] etcdctl stderr:
INFO[0000] Storing upgrade ID on machine                 machine=default/target-cluster02-controlplane-0 upgrade-id=test01
INFO[0000] Checking etcd health                          machine=default/target-cluster02-controlplane-0 upgrade-id=test01
INFO[0000] Running etcdctl                               args="member list -w json" pod=etcd-target-cluster02-controlplane-0
INFO[0000] etcdctl stdout: {"header":{"cluster_id":9521744471245018578,"member_id":7742574140227613697,"raft_term":2},"members":[{"ID":7742574140227613697,"name":"target-cluster02-controlplane-0","peerURLs":["https://192.168.3.98:2380"],"clientURLs":["https://192.168.3.98:2379"]}]}
INFO[0000] etcdctl stderr:
INFO[0000] Running etcdctl                               args="endpoint health --endpoints https://192.168.3.98:2379" pod=etcd-target-cluster02-controlplane-0
INFO[0000] etcdctl stdout: https://192.168.3.98:2379 is healthy: successfully committed proposal: took = 761.531µs
https://192.168.3.98:2379 is healthy: successfully committed proposal: took = 560.745µs
INFO[0000] etcdctl stderr:
INFO[0000] Updating infrastructure reference             api-version=infrastructure.cluster.x-k8s.io/v1alpha2 kind=VSphereMachine machine=default/target-cluster02-controlplane-0 name=target-cluster02-controlplane-0 upgrade-id=test01
INFO[0000] Updating bootstrap reference                  api-version=bootstrap.cluster.x-k8s.io/v1alpha2 kind=KubeadmConfig machine=default/target-cluster02-controlplane-0 name=target-cluster02-controlplane-0 upgrade-id=test01
Error: KubeadmConfig.bootstrap.cluster.x-k8s.io "target-cluster02-controlplane-0.upgrade.test01" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"bootstrap.cluster.x-k8s.io/v1alpha2", "kind":"KubeadmConfig", "metadata":map[string]interface {}{"annotations":map[string]interface {}{"kubectl.kubernetes.io/last-applied-configuration":"{\"apiVersion\":\"bootstrap.cluster.x-k8s.io/v1alpha2\",\"kind\":\"KubeadmConfig\",\"metadata\":{\"annotations\":{},\"name\":\"target-cluster02-controlplane-0\",\"namespace\":\"default\"},\"spec\":{\"clusterConfiguration\":{\"apiServer\":{\"extraArgs\":{\"cloud-provider\":\"external\"}},\"controllerManager\":{\"extraArgs\":{\"cloud-provider\":\"external\"}},\"imageRepository\":\"k8s.gcr.io\"},\"initConfiguration\":{\"nodeRegistration\":{\"criSocket\":\"/var/run/containerd/containerd.sock\",\"kubeletExtraArgs\":{\"cloud-provider\":\"external\"},\"name\":\"{{ ds.meta_data.hostname }}\"}},\"preKubeadmCommands\":[\"hostname \\\"{{ 
....
....
....
validation failure list:
spec.users.passwd in body must be of type string: "null"
```

Adding an empty string to `spec.users.passwd` field of the control plane KubeadmConfig fixed it. I thought it was weird that `spec.users.passwd` [doesn't have the omitempty json tag](https://github.com/kubernetes-sigs/cluster-api/blob/master/bootstrap/kubeadm/api/v1alpha2/kubeadmbootstrapconfig_types.go#L171). I think that's the correct fix but I don't have history/context on that.

**Comment by ncdc** [comment]
We probably need omitempty as you discovered. cc @chuckha

**Comment by detiber** [comment]
I suspect this hasn't been an issue previously because CAPA does not set `Spec.Users` by default.
